### ISSUE TITLE
External checks not thread-safe

### ISSUE THREAD

**Issue by wtarreau** [haproxy/haproxy#141]
While reviewing the external checks code, it appeared that this code is not thread-safe. For example, another thread might be calling malloc() at the moment of the fork(), resulting in one of malloc()'s internal locks to appear as still held in the child process, making extchk_setenv() block forever.

Most likely a good approach would be to call thread_isolate() before the fork() and thread_release() before returning. There is not need to call it in the child (and must not be done as we may block waiting for another thread to release it). We'd be certain to be alone. The extra cost compared to a fork+exec is negligible.

**Comment by lukastribus** [comment]
Likely related:

https://www.mail-archive.com/haproxy@formilux.org/msg34376.html

**Comment by jaroslawr** [comment]
@wtarreau We seem to be running into this problem, which manifests itself by accumulating defunct processes. We tested the patch from the HAProxy mailing list, the one with thread_isolate(), but this resulted in high load on the servers.

Do you think it would be possible to avoid this issue by not doing any malloc() after the fork? Could extchk_setenv() be done before the fork(), from the parent process ?

**Comment by wtarreau** [comment]
It is not expected that thread_isolate() causes a high load, but maybe
you have a high check rate and in this case I can imagine how this could
become expensive.

I don't know much what could be done for this as I don't know this code
well. If you have ideas such as moving some code before the fork(), feel
free to experiment, report and suggest. Maybe some parts will not be
acceptable, maybe they will. In any case this could help design a workable
solution.

**Comment by yuzhichang** [comment]
I'm new to haproxy code. thread_isolate() and thread_release() look like a global mutex impl. So why not use pthread_mutex or pthread_spinlock directly?

**Comment by yuzhichang** [comment]
@wtarreau I tried your thread_isolate patch, days later there are above 80K defunct processes.

**Comment by wtarreau** [comment]
@yuzhichang if you have 80k defunct processes it means that some SIGCHLD is not collecting them, which is a minor issue and should be easy to fix, but more importantly it probably indicates there are multiple issues with the external checks. Any chance you can reduce your config to a very simple reproducer that would help us try to fix this ?

**Comment by wtarreau** [comment]
Well after digging a bit more into the code, I really don't see how it's possible to run into this situation, because as soon as one SIGCHLD strikes, all defunct processes are collected. Thus this means your SIGCHLD is totally blocked, which normally is not the case. I've seen that SIGCHLD is blocked on the thread dealing with the fork() during this operation to avoid a race, but it's immediately unblocked and for me even if I add a delay in the middle, the defuncts are collected.

However I've seen something particularly strange: I'm observing a different behaviour when the checks run on thread 0 and on other threads. On thread 0 I never see the "kill %d" messages but I don't see the defuncts in the background either. I do see this only on other threads. From the debugging I've installed, I'm seeing that elem->exited==1 on thread0 while it's 0 on other threads. I suspect the pid collection is totally racy and probably happens on different threads than the ones issuing the checks. So it's even better, the sigchld handler is never called for processes created on other threads than the first one. Maybe it's only registered on the first thread and the SIGCHLD are directed to specific threads. And maybe in your case it's always the same thread which issues the checks, resulting in SIGCHLD never being seen. I still have to dig a bit more. I always hated this feature because I knew that the principle was fundamentally flawed, I should have refused it :-(

**Comment by wtarreau** [comment]
So it looks like the registered sig handlers are not inherited by the threads, which used never to be a problem since they were there only for basic maintenance stuff. But with directed signals it doesn't work. It's even mentioned in man pthreads(7) that only the same thread may wait() for the children under linux:
       -  When  one  thread  creates  a  new  child process using fork(2), any
          thread should be able to wait(2) on the child.  However, the  imple-
          mentation  allows  only the thread that created the child to wait(2)
          on it.

Thus we're screwed. What I can propose is to force that only the first thread runs the external checks. Anyway we don't need to load-balance such check across all threads, as there must really not be a high load when doing this or there's a bigger issue on the system!

Please try the attached patch which forces the first thread only to run the checks. It works for me (TM). You'll still need the first one with thread_isolate() to avoid the thread safety issues I mentioned but these ones are expected to be rare enough so that if you omit it from your first test it should not cause a big trouble.
[fix2.diff.txt](https://github.com/haproxy/haproxy/files/3570565/fix2.diff.txt)

**Comment by wtarreau** [comment]
Sorry there was another place, the one above is not sufficient, please use this one instead.
[fix3.diff.txt](https://github.com/haproxy/haproxy/files/3570608/fix3.diff.txt)

**Comment by yuzhichang** [comment]
@wtarreau With your fix3.diff.txt, haproxy runs for 2 weeks without accumulating defunct processes, and external-checks work well.

**Comment by wtarreau** [comment]
Excellent, many thanks for confirming! It was already merged into mainline and backported to 2.0.

**Comment by wtarreau** [comment]
Was fixed as 6dd4ac8 and backported to 1.8, closing now.
